chore(deps): Update dependencies

reneleonhardt opened the pull request:
### Changes
- Update Python dependencies (only `lint` and `test` had updates)
- Update Cargo dependencies (`mimalloc` and `tokio`)
- Let Dependabot update Cargo and Python dependencies weekly
- Remove unused `_pgo_deps.txt` (no PGO builds since 2.0)
- Add badges to README.md (I can drop my improvements if you don't want them)
![PyPI - Version](https://img.shields.io/pypi/v/granian)
![PyPI - Downloads](https://img.shields.io/pypi/dm/granian)
[![Build Status](https://github.com/emmett-framework/granian/actions/workflows/build.yml/badge.svg)](https://github.com/emmett-framework/granian/actions)
![Python Version from PEP 621 TOML](https://img.shields.io/python/required-version-toml?tomlFilePath=https://raw.githubusercontent.com/emmett-framework/granian/master/pyproject.toml)
![PyPI - License](https://img.shields.io/pypi/l/granian)

### Notes
- While Microsoft is taking their time, a preliminary patch has been released
  https://github.com/purpleprotocol/mimalloc_rust/issues/133#issuecomment-2784431691
  Are there Alpine tests or can you check the build manually?
- I guess the mismatch between cargo.toml and .lock confuses Dependabot 😅
  https://github.com/emmett-framework/granian/pull/449
  So I resynced them, 58 tests passed with tokio 1.44.2, hopefully Dependabot can work again.
- There are no PGO builds anymore, can the corresponding test cases be removed? 
- Was there a reason to pin Python to 3.11 in benchmarks or can it be updated to 3.12 or 3.13?
- [PyPy 3.9 has been discontinued](https://pypy.org/posts/2024/08/pypy-v7317-release.html) in August, can it be removed?

gi0baro reviewed (changes requested):
Please revert all the changes that doesn't resemble the title of your PR. Specifically:
- `.github/dependabot.yml`
- `README.md`

Motivation and rationale:
- Updating dependencies is one thing, changing workflows/README is a totally different thing and requires relative discussions
- There's no need to bother maintainers once a week, especially considering that every rust dependency change would require a new release to be effective and all the main Python dependencies have *relaxed* pins
- I don't see any valid reason to bloat the README with all that

Also there's no point in updating `Cargo.toml`, we don't need to require a minimum version for `tokio` and `mimalloc` different from what it's already there.

Regarding your questions:
> Are there Alpine tests or can you check the build manually?

I can manually trigger a build once this is ready

> I guess the mismatch between cargo.toml and .lock confuses Dependabot 😅
https://github.com/emmett-framework/granian/pull/449
> So I resynced them, 58 tests passed with tokio 1.44.2, hopefully Dependabot can work again.

I'm not sure what you're talking about, the version that matters is the one in Cargo.lock, current release is already using tokio `1.44.2`.

> [PyPy 3.9 has been discontinued](https://pypy.org/posts/2024/08/pypy-v7317-release.html) in August, can it be removed?

Since we still support Python 3.9, I can't see any specific reason to drop wheels for PyPy 3.9.

reneleonhardt commented:
> Please revert all the changes that doesn't resemble the title of your PR. Specifically:
> 
>     * `.github/dependabot.yml`
> 
>     * `README.md`

You can just close this PR if there are no improvements in your eyes.
I removed dependabot and README, but of course there was a reason to let Dependabot update pyproject.toml automatically instead of waiting for contributors to do it manually. In addition, I fixed your old GitHub link as you saw. And weekly could be changed to monthly in a second if that's a problem... 
The README also contained improvements regarding development, but I'm sure you saw that.

> Motivation and rationale:
> 
>     * Updating dependencies is one thing, changing workflows/README is a totally different thing and requires relative discussions

I didn't change workflows.
I even changed the README in a separate commit, it was just a suggestion and as maintainer you can change everything in my PR yourself anyway if it is totally against your principles 🤷

>     * There's no need to bother maintainers once a week, especially considering that every rust dependency change would require a new release to be effective and all the main Python dependencies have _relaxed_ pins
Dependabot doesn't release itself, it just creates PRs and your workflows check if everything is still working.
You can just leave those PRs open if you don't agree with Dependabot that they are worth releasing?
In addition, you can group dependencies to reduce the number of PRs if the amount of weekly updates is bothering you.

>     * I don't see any valid reason to bloat the README with all that
It's a valid opinion.
On the other hand many developers see badges in a positive way to inform their users and contributors in the blink of an eye.

> Also there's no point in updating `Cargo.toml`, we don't need to require a minimum version for `tokio` and `mimalloc` different from what it's already there.

I reverted Cargo.toml. FYI cargo-upgrade changed both files.

> Regarding your questions:
> 
> > Are there Alpine tests or can you check the build manually?
> 
> I can manually trigger a build once this is ready
> 
> > I guess the mismatch between cargo.toml and .lock confuses Dependabot 😅
> > #449
> > So I resynced them, 58 tests passed with tokio 1.44.2, hopefully Dependabot can work again.
> 
> I'm not sure what you're talking about, the version that matters is the one in Cargo.lock, current release is already using tokio `1.44.2`.
I confused myself, cargo-upgrade didn't change tokio in lock, only in toml, it seems letting them stay in sync makes it easier for tools too 😅

> > [PyPy 3.9 has been discontinued](https://pypy.org/posts/2024/08/pypy-v7317-release.html) in August, can it be removed?
> 
> Since we still support Python 3.9, I can't see any specific reason to drop wheels for PyPy 3.9.
Yes. Except... that pypy isn't covered in test.yml. So can the build workflow be sure that pypy wheels work? 🤔
